src/fullnode.rs

```diff
@@ -382,6 +425,38 @@ impl Fullnode {
         }
     }
 
+    // Runs a thread to manage node role transitions.  The returned closure can be used to signal the
+    // node to exit.
+    pub fn run(
+        mut self,
+        rotation_notifier: Option<Sender<(FullnodeReturnType, u64)>>,
+    ) -> impl FnOnce() {
+        let (sender, receiver) = channel();
+        let exit = self.exit.clone();
+        spawn(move || loop {
+            let status = self.handle_role_transition();
+            match status {
+                None => {
+                    debug!("node shutdown requested");
+                    self.close().expect("Unable to close node");
+                    sender.send(true).expect("Unable to signal exit");
+                    break;
+                }
+                Some(transition) => {
+                    debug!("role_transition complete: {:?}", transition);
+                    if let Some(ref rotation_notifier) = rotation_notifier {
+                        rotation_notifier.send(transition).unwrap();
+                    }
+                }
+            };
+        });
+        move || {
```

Comment:
this is clever, what happens on drop() of the closure?

Reply:
If/when somebody else sets `exit` to true eventually the thread above should panic at `sender.send(true).expect("Unable to signal exit");`, which seams fine cause it's done working at that point anyway